drivers/hmc5883l/hmc5883l.c

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (C) 2018 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_hmc5883l
+ * @brief       Device driver for the Honeywell HMC5883L 3-axis digital compass
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @file
+ * @{
+ */
+
+#include <string.h>
+#include <stdlib.h>
+
+#include "hmc5883l_regs.h"
+#include "hmc5883l.h"
+
+#include "log.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+
+#define DEBUG_DEV(f, d, ...) \
+        DEBUG("[hmc5883l] %s i2c dev=%d addr=%02x: " f "\n", \
+              __func__, d->params.dev, d->params.addr, ## __VA_ARGS__);
+
+#else /* ENABLE_DEBUG */
+
+#define DEBUG_DEV(f, d, ...)
+
+#endif /* ENABLE_DEBUG */
+
+#define ERROR_DEV(f, d, ...) \
+        LOG_ERROR("[hmc5883l] %s i2c dev=%d addr=%02x: " f "\n", \
+                  __func__, d->params.dev, d->params.addr, ## __VA_ARGS__);
+
+#define EXEC_RET(f, r) \
+    if ((r = f) != HMC5883L_OK) { \
+        DEBUG("[hmc5883l] %s: error code %d\n", __func__, res); \
+        return res; \
+    }
+
+#define EXEC_RET_CODE(f, r, c) \
+    if ((r = f) != HMC5883L_OK) { \
+        DEBUG("[hmc5883l] %s: error code %d\n", __func__, res); \
+        return c; \
+    }
+
+/** Forward declaration of functions for internal use */
+
+static int _is_available(const hmc5883l_t *dev);
+
+static int _reg_read(const hmc5883l_t *dev, uint8_t reg, uint8_t *data, uint16_t len);
+static int _reg_write(const hmc5883l_t *dev, uint8_t reg, uint8_t data);
+
+int hmc5883l_init(hmc5883l_t *dev, const hmc5883l_params_t *params)
+{
+    int res = HMC5883L_OK;
+
+    assert(dev != NULL);
+    assert(params != NULL);
+    DEBUG_DEV("params=%p", dev, params);
+
+    /* init sensor data structure */
+    dev->params = *params;
+
+    /* check availability of the sensor */
+    EXEC_RET(_is_available(dev), res)
+
+    uint8_t cfg_a = 0;
+
+    /* set configuration register A and B */
+    cfg_a = params->meas_avg | params->meas_mode | params->dor;
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_CFG_A, cfg_a), res);
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_CFG_B, params->gain), res);
```

Comment:
There is quite some unneeded overhead introduced here by aquiring/releasing the i2c bus for every singe byte transferred. It would be more elegant to acquire once, do the needed transfers and release after...

Reply:
Hm, expanding `_reg_write` to `i2c_*` primtives would let explode the source code. Could you live with something like that
```
    uint8_t data[3] = { params->meas_avg | params->meas_mode | params->dor,
                        params->gain, 
                        params->op_mode };
    EXEC_RET(_reg_write(dev, HMC5883L_REG_CFG_A, data, 3), res);
```
that uses the fact that the registers are in subsequent order?

Regarding the code size, the former implementation required
```
   text	   data	    bss	    dec	    hex
  13148	    424	    953	  14525	   38bd
```
This implementation would require less memory.
```
   text	   data	    bss	    dec	    hex
  13130	    424	    953	  14507	   
```

Reply:
The implementation of `_reg_write` and `_reg_read` with `acquire` and `release` as in #12464 requires much more memory:
```
   text	   data	    bss	    dec	    hex	filename
  13202	    424	    953	  14579	   38f3
```

Reply:
So, I would tend to leave my implementation as it is. The only place where the release in between subsequent register accesses happens is in `hmc5883_init`.